core/templates/services/platform-feature.service.ts

```diff
@@ -0,0 +1,285 @@
+import { WindowRef } from 'services/contextual/window-ref.service';
+
+interface FeatureFlagsCacheItem {
+  timestamp: number;
+  sessionId: string;
+  featureStatusSummary: FeatureStatusSummary;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class PlatformFeatureService {
+  private static SESSION_STORAGE_KEY = 'SAVED_FEATURE_FLAGS';
+  private static SESSION_STORAGE_CACHE_TTL = 12 * 3600 * 1000; // 12 hours
+
+  private static COOKIE_NAME_FOR_SESSION_ID = 'SACSID';
+  private static COOKIE_NAME_FOR_SESSION_ID_IN_DEV = 'dev_appserver_login';
+
+  // TODO(#9154): Remove static when migration is complete.
+  static featureStatusSummary: FeatureStatusSummary = null;
+
+  // TODO(#9154): Remove static when migration is complete.
+  static _initializedWithError = false;
+
+  // TODO(#9154): Remove static when migration is complete.
+  static initializationPromise: Promise<void> = null;
+
+  constructor(
+      private clientContextObjectFactory: ClientContextObjectFactory,
+      private platformFeatureBackendApiService:
+        PlatformFeatureBackendApiService,
+      private featureStatusSummaryObjectFactory:
+        FeatureStatusSummaryObjectFactory,
+      private i18nLanguageCodeService: I18nLanguageCodeService,
+      private windowRef: WindowRef) {
+    this.initialize();
+  }
+
+  /**
+   * Inializes the PlatformFeatureService, it's guaranteed that it's only
+   * inialized once.
+   *
+   * @returns {Promise} - A promise that is resolved when the initialization
+   * is done.
+   */
+  async initialize(): Promise<void> {
+    if (!PlatformFeatureService.initializationPromise) {
+      PlatformFeatureService.initializationPromise = this._initialize();
+    }
+    return PlatformFeatureService.initializationPromise;
+  }
+
+  /**
+   * Returns the summary object of feature flags, which can be used to get the
+   * value of feature flags.
+   * Example:
+   *   platformFeatureService.featureSummary.DummyFeature.isEnabled === true.
+   *
+   * @returns {FeatureSummaryDict} - Summary object of feature flags.
+   * @throws {Error} - If this method is called before inialization.
+   */
+  get featureSummary(): FeatureSummaryDict {
+    if (PlatformFeatureService.featureStatusSummary) {
+      return PlatformFeatureService.featureStatusSummary.toSummaryDict();
+    } else {
+      throw new Error('The platform feature service has not been initialized.');
+    }
+  }
+
+  /**
+   * Checks if there's any error, e.g. request timeout, during initialization.
+   *
+   * @returns {boolean} - True if there is any error during initialization.
+   */
+  get initialzedWithError(): boolean {
+    return PlatformFeatureService._initializedWithError;
+  }
+
+  /**
+   * Detects the type of browser from it's userAgent.
+   *
+   * @returns {string} - The name of the browser that is being used.
+   */
+  detectBrowserType(): string {
+    const ua = this.windowRef.nativeWindow.navigator.userAgent;
+
+    if (ua.includes('edg') || ua.includes('Edge')) {
```

Comment:
Suggest adding a comment with a link to that article to provide context since others might assume this is a typo, too.

Also, I suggest doing the Edge check last. In terms of likelihood of the browser name accidentally showing up in the user agent, 'edge' is a much more common word than the order browser names per random English.

Reply:
Update: The ordering of the browser check has to be in the original way, because the user agent of Edge also includes `Chrome`.

Reply:
Ah, darn. Nice catch. Probably worth adding a brief comment here, then, that the ordering is important to avoid accidentally mixing up browsers.

Reply:
Done, added comment for the ordering.